File: Assets/code/DD.java
import java.io.*;

class DD {

    static class Node {
        int data;
        Node prev;
        Node next;

        public Node(int data) {
            this.data = data;
            this.prev = null;
            this.next = null;
        }
    }

    static Node deleteFirstNode(Node head) {
        if (head == null) {
            System.out.println("List is empty. Unable to delete.");
            return null;
        }
        Node temp = head;
        head = head.next;
        if (head != null) {
            head.prev = null;
        }
        temp.next = null;
        System.out.println("First node deleted successfully.");
        return head;
    }

    static Node deleteLastNode(Node head) {
        if (head == null) {
            System.out.println("List is empty. Unable to delete.");
            return null;
        }
        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }
        if (temp.prev != null) {
            temp.prev.next = null;
        } else {
            head = null;
        }
        temp.prev = null;
        System.out.println("Last node deleted successfully.");
        return head;
    }

    static Node deleteNode(Node head, int key) {
        if (head == null) {
            System.out.println("List is empty. Unable to delete.");
            return null;
        }
        Node temp = head;
        while (temp != null && temp.data != key) {
            temp = temp.next;
        }
        if (temp == null) {
            System.out.println("Node with data " + key + " not found.");
        } else {
            if (temp.prev != null) {
                temp.prev.next = temp.next;
            } else {
                head = temp.next;
            }
            if (temp.next != null) {
                temp.next.prev = temp.prev;
            }
            temp.prev = null;
            temp.next = null;
            System.out.println("Node with data " + key + " deleted successfully.");
        }
        return head;
    }

    static void displayDoublyLinkedList(Node head) {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Node head = new Node(1);
        Node second = new Node(2);
        Node third = new Node(3);

        head.next = second;
        second.prev = head;
        second.next = third;
        third.prev = second;

        System.out.print("Original Doubly Linked List: ");
        displayDoublyLinkedList(head);

        head = deleteFirstNode(head);
        displayDoublyLinkedList(head);

        head = deleteLastNode(head);
        displayDoublyLinkedList(head);

        head = deleteNode(head, 2);
        displayDoublyLinkedList(head);
    }
}

File: Assets/code/DD.c

#include<stdio.h> 
#include<stdbool.h>

// Structure for a node in the doubly linked list
struct Node {
    int data;
    struct Node* prev;
    struct Node* next;
};

// Function to delete the first node from the linked list
struct Node* deleteFirstNode(struct Node* head) {
    if (head == NULL) {
        printf("List is empty. Unable to delete.\n");
        return NULL;
    }
    struct Node* temp = head;
    head = head->next;
    if (head != NULL) {
        head->prev = NULL;
    }
    free(temp);
    printf("First node deleted successfully.\n");
    return head;
}

// Function to delete the last node from the linked list
struct Node* deleteLastNode(struct Node* head) {
    if (head == NULL) {
        printf("List is empty. Unable to delete.\n");
        return NULL;
    }
    struct Node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    if (temp->prev != NULL) {
        temp->prev->next = NULL;
    } else {
        head = NULL;
    }
    free(temp);
    printf("Last node deleted successfully.\n");
    return head;
}

// Function to delete a node with a given key from the linked list
struct Node* deleteNode(struct Node* head, int key) {
    if (head == NULL) {
        printf("List is empty. Unable to delete.\n");
        return NULL;
    }
    struct Node* temp = head;
    while (temp != NULL && temp->data != key) {
        temp = temp->next;
    }
    if (temp == NULL) {
        printf("Node with data %d not found.\n", key);
    } else {
        if (temp->prev != NULL) {
            temp->prev->next = temp->next;
        } else {
            head = temp->next;
        }
        if (temp->next != NULL) {
            temp->next->prev = temp->prev;
        }
        free(temp);
        printf("Node with data %d deleted successfully.\n", key);
    }
    return head;
}

// Function to display the doubly linked list
void displayDoublyLinkedList(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

int main() {
    // Create and initialize nodes for the doubly linked list
    struct Node* head = (struct Node*)malloc(sizeof(struct Node));
    struct Node* second = (struct Node*)malloc(sizeof(struct Node));
    struct Node* third = (struct Node*)malloc(sizeof(struct Node));

    head->data = 1;
    head->prev = NULL;
    head->next = second;

    second->data = 2;
    second->prev = head;
    second->next = third;

    third->data = 3;
    third->prev = second;
    third->next = NULL;

    printf("Original Doubly Linked List: ");
    displayDoublyLinkedList(head);

    // Perform deletion operations
    head = deleteFirstNode(head);
    displayDoublyLinkedList(head);

    head = deleteLastNode(head);
    displayDoublyLinkedList(head);

    head = deleteNode(head, 2);
    displayDoublyLinkedList(head);

    return 0;
}
              
            
